[networking] gather per-namespace ip addresses

robbmanes opened the pull request:
Collecting the addresses in each namespace is beneficial instead of
only the name of the namespace itself.  This patch checks to see if a
namespace exists, and if it does, performs an:

```
$ ip netns exec <namespace> ip address show
```

Therefore, capturing the addresses in each namespace.

Signed-off-by: Robb Manes <rmanes@redhat.com>

robbmanes commented:
Forgive me if I failed to submit this right so that it can be applied via fast-forward, I'd be happy to re-do.

Submitting a patch for the collection of ip netns details, notably the ip
addresses contained in each.  This is useful for troubleshooting, particularly
using cloud products that rely on namespaces (like OpenStack).

Sosreports using the updated networking plugin can be found at:

    http://people.redhat.com/rmanes/sos-patch/

I was concerned about the amount of time it would take to do this per namespace,
and thus wrote a dumb little script to set up ~1000 namespaces (although empty):

```
    #!/usr/bin/env python

    from subprocess import call
    import sys

    def create_test_namespace(name):
    	call(["ip", "netns", "add", str(name)])

    def main():
    	for i in range(0, 1000):
    		create_test_namespace(("test_namespace" + str(i)))
    	return 0

    if __name__ == "__main__":
    	sys.exit(main())


    # ./sosreport -o networking
    - - - - - - - - - 8< - - - - - - - - - 
    real	0m40.115s
    user	0m2.082s
    sys	0m19.686s
```

All unit tests appear to have passed:

```
    $ nosetests
    ..................................................................................
    ----------------------------------------------------------------------
    Ran 82 tests in 0.861s

    OK
```

robbmanes commented:
Closing and will re-open after I apply some minor fixes (pep8 and spelling, somehow this passed on my local system and I have no idea why).

robbmanes commented:
Reopening with fixed pep8 and spelling; if there's any other problems let me know and I'd be happy to take care of them.

battlemidget commented:
ack, thanks

bmr-cymru commented:
Please squash the pep8 fixups into the original patch - we try to keep the tree passing the test suite on each commit so that bisects are easy.

Generally though, could you take a look at Issue #256 and see if there isn't a way to move a bit closer to that goal with these changes?

Right now we have two plugins handling network data and net namespaces which is a bit crazy. The Openstack bits were rushed in to support their use cases but we really should be trying to consolidate all this rather than adding on more and more bits in a piecemeal fashion to both.

robbmanes commented:
@bmr-cymru Ah, I didn't even see that issue.  Yes, let me take a look at that and see if I can help there; the instance in which I personally needed this was not OpenStack-specific.

I'll also fix the original patch.  Thanks!

bmr-cymru commented:
Thanks! If it seems unworkable for now then I am not completely opposed to this change - but if you think we can start moving toward a more coherent approach for this (and have the netns knowledge to implement it) then that's something I'd really like to see.

robbmanes commented:
@bmr-cymru After reading the issue, I'm of the opinion that we should keep the namespace-specific sections in the networking plugin, and glob dhcp files/settings/commands into neutron as neutron acts as the network management layer here, but isn't necessarily the only one that would use namespaces (although it is, as of this moment, one of the most prevalent).

Let me expand on my patch to get more out of namespaces like routing and netfilter, in a way that neutron and/or any other namespace cognizant application would benefit from collecting besides just neutron; I'll update soon after I test it working on some nodes.  Then we can work on a separate patch for neutron-specifics, if that sounds good to you?

bmr-cymru commented:
Sounds like a plan - thanks for taking this on!

robbmanes commented:
Mistaken command, sorry.

robbmanes commented:
Updated my branch with my thought process, but I'm still testing.  I still have to put it through an intensive namespace test with multiple interfaces, which I'll do once I'm sure everything works. I'll let you know if all goes as expected.

I've taken a lot out of the aforementioned issue and moved it into the networking plugin, as it's all perfectly valid data to capture that isn't necessarily specific to neutron.  Thoughts, comments, criticism is welcome.

robbmanes commented:
Okay; as of the moment all tests pass and we collect the data we need.  The complexity from my original has gone up a bunch, so, feel free to test in any way you can conceive of.

I have tested this on a few neutron nodes along with several instances of creating namespaces and veth-pairs to verify that we capture the routing information along with the ethtool output.  Thusfar, all looks as expected, even when I assign strange names (including the `@NONE` that appears to unbound/partnerless veth peer devices).

Example from neutron:
```
$ ll sos_commands/networking | grep ip_netns | head
-rw-r--r--. 1 root root   1107 Sep 11 17:37 ip_netns
-rw-r--r--. 1 root root    286 Sep 11 17:38 ip_netns_exec_qdhcp-2104a085-63a3-4e81-867f-acc3567b40d1_ip_address_show
-rw-r--r--. 1 root root    553 Sep 11 17:38 ip_netns_exec_qdhcp-2104a085-63a3-4e81-867f-acc3567b40d1_ip_route_show_table_all
-rw-r--r--. 1 root root    822 Sep 11 17:38 ip_netns_exec_qdhcp-2104a085-63a3-4e81-867f-acc3567b40d1_iptables-save
-rw-r--r--. 1 root root    286 Sep 11 17:38 ip_netns_exec_qdhcp-3b1fc0f5-6c42-422c-a9b7-7abe913cd5bc_ip_address_show
-rw-r--r--. 1 root root    553 Sep 11 17:38 ip_netns_exec_qdhcp-3b1fc0f5-6c42-422c-a9b7-7abe913cd5bc_ip_route_show_table_all
-rw-r--r--. 1 root root    819 Sep 11 17:38 ip_netns_exec_qdhcp-3b1fc0f5-6c42-422c-a9b7-7abe913cd5bc_iptables-save
-rw-r--r--. 1 root root    286 Sep 11 17:38 ip_netns_exec_qdhcp-476c2d1e-b3c8-4b9e-8a83-81d788414e63_ip_address_show
-rw-r--r--. 1 root root    553 Sep 11 17:38 ip_netns_exec_qdhcp-476c2d1e-b3c8-4b9e-8a83-81d788414e63_ip_route_show_table_all
-rw-r--r--. 1 root root    795 Sep 11 17:38 ip_netns_exec_qdhcp-476c2d1e-b3c8-4b9e-8a83-81d788414e63_iptables-save
```

Test sosreport from a working, active neutron node available at:
http://people.redhat.com/rmanes/sos-patch/sosreport-neutron-test.tar.bz2

bmr-cymru commented:
Looks good - I'll try to do some testing myself in the next few days but I the changes all seem sane.

lyarwood commented:
:+1: LGTM, ran it locally on an OSP-d deployed RHEL OSP 7 deployment and everything appeared to be correct.

battlemidget commented:
ACK based on previous ack's and @lyarwood testing
